internal/util/function/rerank/model_function.go

```diff
@@ -0,0 +1,384 @@
+func newProvider(params []*commonpb.KeyValuePair) (modelProvider, error) {
+	for _, param := range params {
+		if strings.ToLower(param.Key) == providerParamName {
+			provider := strings.ToLower(param.Value)
+			conf := paramtable.Get().FunctionCfg.GetRerankModelProviders(provider)
+			switch provider {
+			case vllmProviderName:
+				return newVllmProvider(params, conf)
+			case teiProviderName:
+				return newTeiProvider(params, conf)
+			default:
+				return nil, fmt.Errorf("Unknow rerank provider:%s", param.Value)
+			}
+		}
+	}
+	return nil, fmt.Errorf("Lost rerank params:%s ", providerParamName)
+}
+
+type ModelFunction[T PKType] struct {
+	RerankBase
+
+	provider modelProvider
+	queries  []string
+}
+
+func newModelFunction(collSchema *schemapb.CollectionSchema, funcSchema *schemapb.FunctionSchema) (Reranker, error) {
+	base, err := newRerankBase(collSchema, funcSchema, decayFunctionName, false)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(base.GetInputFieldNames()) != 1 {
+		return nil, fmt.Errorf("Rerank model only supports single input, but gets [%s] input", base.GetInputFieldNames())
+	}
+
+	inputType := schemapb.DataType_None
+	for _, field := range collSchema.Fields {
+		if field.Name == base.GetInputFieldNames()[0] {
+			inputType = field.DataType
+		}
+	}
+	if inputType != schemapb.DataType_VarChar {
+		return nil, fmt.Errorf("Rerank model only support varchar, bug got [%s]", inputType.String())
+	}
+
+	provider, err := newProvider(funcSchema.Params)
+	if err != nil {
+		return nil, err
+	}
+
+	queries := []string{}
+	for _, param := range funcSchema.Params {
+		if param.Key == queryKeyName {
+			if err := json.Unmarshal([]byte(param.Value), &queries); err != nil {
+				return nil, fmt.Errorf("Parse rerank params [queries] failed, err: %v", err)
+			}
+		}
+	}
+	if len(queries) == 0 {
+		return nil, fmt.Errorf("Rerank function lost params queries")
+	}
+
+	if base.pkType == schemapb.DataType_Int64 {
+		return &ModelFunction[int64]{RerankBase: *base, provider: provider, queries: queries}, nil
+	} else {
+		return &ModelFunction[string]{RerankBase: *base, provider: provider, queries: queries}, nil
+	}
+}
+
+func (model *ModelFunction[T]) processOneSearchData(ctx context.Context, searchParams *SearchParams, query string, cols []*columns) (*IDScores[T], error) {
+	uniqueData := make(map[T]string)
+	for _, col := range cols {
+		texts := col.data[0].([]string)
+		ids := col.ids.([]T)
+		for idx, id := range ids {
+			if _, ok := uniqueData[id]; !ok {
+				uniqueData[id] = texts[idx]
+			}
+		}
+	}
+	ids := make([]T, 0, len(uniqueData))
+	texts := make([]string, 0, len(uniqueData))
+	for id, text := range uniqueData {
+		ids = append(ids, id)
+		texts = append(texts, text)
+	}
+	scores, err := model.provider.rerank(ctx, query, texts)
+	if err != nil {
+		return nil, err
+	}
+
+	rerankScores := map[T]float32{}
+	for idx, id := range ids {
+		rerankScores[id] = scores[idx]
+	}
+	return newIDScores(rerankScores, searchParams), nil
```

Comment:
So we only do rescore instead of reranking?

Reply:
The rerank model re-scores based on the query and doc content